### PR TITLE
ch4/ofi: RMA key allocator rework.

### PR DESCRIPTION
**Pull request by alexander-sannikov**
- RMA-key allocator is implemented as a global entity
- 64-bit keys support added into key allocator
- Simplified RMA-kay layout

### CODE REVIEW

**Comment by alexander-sannikov** [comment]
@yfguo I've removed unnecessary entities and context_id dependency. What do you think about? Also we should protect index allocator with a lock I suppose.

File: src/mpid/ch4/netmod/ofi/ofi_init.h [hunk 1]
```diff
@@ -608,20 +608,11 @@ static inline int MPIDI_NM_mpi_init_hook(int rank,
...
-        MPIDI_Global.max_huge_rmas = MPIDI_OFI_MAX_HUGE_RMAS_32;
-        MPIDI_Global.context_shift = MPIDI_OFI_CONTEXT_SHIFT_32;
-        MPIDI_Global.rma_key_type_bits = MPIDI_OFI_MAX_KEY_TYPE_BITS_32;
+        MPIDI_Global.max_mr_key_bits = MPIDI_OFI_MAX_KEY_BITS_32;
     } else if (MPIDI_Global.max_mr_key_size >= 2) {
-        MPIDI_Global.max_rma_key_bits = MPIDI_OFI_MAX_KEY_BITS_16;
-        MPIDI_Global.max_huge_rmas = MPIDI_OFI_MAX_HUGE_RMAS_16;
-        MPIDI_Global.context_shift = MPIDI_OFI_CONTEXT_SHIFT_16;
-        MPIDI_Global.rma_key_type_bits = MPIDI_OFI_MAX_KEY_TYPE_BITS_16;
+        MPIDI_Global.max_mr_key_bits = MPIDI_OFI_MAX_KEY_BITS_16;
```
Comment (pavanbalaji) [inline src/mpid/ch4/netmod/ofi/ofi_init.h]: Why special case these just for 8, 4, 2 bytes?  Why not set `MPIDI_Global.max_mr_key_bits` to `MPIDI_Global.max_mr_key_size * 8`?

**Comment by raffenet** [comment]
test:jenkins/ch4/ofi

File: src/mpid/ch4/netmod/ofi/ofi_init.h (hunk 1 above)
Comment (alexander-sannikov) [reply src/mpid/ch4/netmod/ofi/ofi_init.h]: Yes, in current state this code may be removed as well.

File: src/mpid/ch4/netmod/ofi/ofi_types.h [hunk 2]
```diff
@@ -388,6 +336,7 @@ typedef struct {
     /* Window/RMA Globals */
     void *win_map;
     uint64_t rma_issued_cntr;
+    void *win_id_allocator;
```
Comment (hajimefu) [inline src/mpid/ch4/netmod/ofi/ofi_types.h]: I feel like this should now be called `mr_key_allocator`.

Comment (hajimefu) [reply src/mpid/ch4/netmod/ofi/ofi_types.h]: Also, if I understood correctly, this single allocator is now used for 
- generating window ID for RMA (=equal to MR key for RMA)
- generating MR key for pt2pt LMT

These design choices have to be written somewhere, probably here.

File: src/mpid/ch4/netmod/ofi/ofi_init.h [hunk 3]
```diff
@@ -911,6 +896,9 @@ static inline int MPIDI_NM_mpi_init_hook(int rank,
     OPA_store_int(&MPIDI_Global.am_inflight_inject_emus, 0);
     OPA_store_int(&MPIDI_Global.am_inflight_rma_send_mrs, 0);
 
+    /* Create RMA keys allocators */
```
Comment (hajimefu) [inline src/mpid/ch4/netmod/ofi/ofi_init.h]: "allocators" -> "allocator"

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi

**Comment by yfguo** [comment]
@alexander-sannikov Hey Alex, can you take a look at the OFI memory registration failure? 
Here is the output of the test of this PR: https://jenkins.mpich.org/job/mpich-review-ch4-ofi/1912/compiler=gnu,jenkins_configure=default,label=ib64,ofi_prov=sockets/console

File: src/mpid/ch4/netmod/ofi/ofi_am_impl.h [hunk 4]
```diff
@@ -297,13 +297,11 @@ static inline int MPIDI_OFI_am_isend_long(int rank,
...
-            MPIDI_OFI_index_allocator_alloc(MPIDI_OFI_COMM
-                                            (MPIR_Process.comm_world).rma_id_allocator,
-                                            MPL_MEM_RMA);
-        MPIR_Assert(index < MPIDI_Global.max_huge_rmas);
-        lmt_info->rma_key = MPIDI_OFI_rma_key_pack(comm->context_id,
-                                                   MPIDI_OFI_KEY_TYPE_HUGE_RMA, index);
+        lmt_info->rma_key =
+            MPIDI_OFI_index_allocator_alloc(MPIDI_Global.mr_key_allocator, MPL_MEM_RMA);
+        MPIR_ERR_CHKANDSTMT(lmt_info->rma_key >=
+                            (((uint64_t) 1) << (MPIDI_Global.max_mr_key_size * 8)), mpi_errno,
```
Comment (yfguo) [inline src/mpid/ch4/netmod/ofi/ofi_am_impl.h]: When `max_mr_key_size` equals 8, ) `((uint64_t) 1) << (MPIDI_Global.max_mr_key_size * 8)` actually equals `1`. Why are we checking if the `rma_key` in this way?

Comment (yfguo) [reply src/mpid/ch4/netmod/ofi/ofi_am_impl.h]: I understand the check if to ensure the `rma_key` is not larger than the OFI MR key range which makes sence, but the case when the OFI MR key allows 64 bits, this check will be incorrect. I think I can fix it.

**Comment by wesbland** [comment]
This work has moved to #3541 and this PR can be closed.

**Pull request closed without merging**